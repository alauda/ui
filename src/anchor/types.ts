import { EventEmitter, TemplateRef } from '@angular/core';

export interface AnchorItemBase {
  readonly id?: string;
  readonly target?: HTMLElement;
  readonly label: TemplateRef<unknown> | string;
  readonly labelContext?: unknown;
  readonly labelChange?: EventEmitter<TemplateRef<unknown> | string>;
}

export interface AnchorItem extends AnchorItemBase {
  readonly level: number;
}

export interface AnchorTreeItem extends AnchorItemBase {
  children?: AnchorTreeItem[];
}
