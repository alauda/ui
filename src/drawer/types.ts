import { ComponentType } from '@angular/cdk/portal';
import { TemplateRef } from '@angular/core';

import { ValueOf } from '../types';

export const DrawerSize = {
  Small: 'small',
  Medium: 'medium',
  Big: 'big',
} as const;

export type DrawerSize = ValueOf<typeof DrawerSize>;

export interface DrawerOptions<T = unknown, C extends object = object> {
  title?: string | TemplateRef<C>;
  content?: ComponentType<T> | TemplateRef<C>;
  footer?: string | TemplateRef<C>;
  contentParams?: C; // 不仅作为content的参数，同时是title和footer的上下文
  width?: number;
  size?: DrawerSize; // 内置的宽度尺寸，也可以使用 width 自定义
  offsetY?: string;
  divider?: boolean;
  drawerClass?: string;
  visible?: boolean;
  showClose?: boolean;
  mask?: boolean;
  maskClosable?: boolean; // 点击背景是否关闭抽屉
  hideOnClickOutside?: boolean; // 在抽屉外点击是否关闭抽屉，与 maskClosable 的区别是是否有 mask
}
