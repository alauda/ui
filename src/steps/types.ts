export enum StepState {
  Default = 'default',
  Done = 'done',
  Error = 'error',
  Pending = 'pending',
}

export interface StepItem {
  label: string;
  description?: string;
  state: StepState;
  optional?: boolean;
  editable?: boolean;
}

export type StepsOrientation = 'horizontal' | 'vertical';

export class StepsSelection {
  selectedIndex: number;
  previousSelectedIndex: number;
  selectedStep: StepItem;
  previousSelectedStep: StepItem;
}
