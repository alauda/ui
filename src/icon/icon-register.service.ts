import { DOCUMENT } from '@angular/common';
import { HttpClient } from '@angular/common/http';
import { Inject, Injectable, Optional, SkipSelf } from '@angular/core';

import { auiIcons } from './icons';
import {
  getAuiIconFailedToLoadCustomIconFile,
  getAuiIconNoHttpProviderError,
} from './utils';

@Injectable()
export class IconRegisterService {
  private defaultIconPrefix = 'aui-icon';
  private readonly doc: Document;

  constructor(
    @Optional()
    @Inject(DOCUMENT)
    document: any,
    @Optional() private readonly httpClient: HttpClient,
  ) {
    this.doc = document;
    this.registerSvgSymbolsByString(auiIcons);
  }

  getDefaultIconPrefix(): string {
    return this.defaultIconPrefix;
  }

  replaceDefaultIconPrefix(prefix: string): void {
    this.defaultIconPrefix = prefix;
  }

  registerSvgSymbolsByUrl(url: string) {
    if (!this.httpClient) {
      throw getAuiIconNoHttpProviderError();
    }
    this.httpClient
      .get(url, {
        responseType: 'text',
      })
      .subscribe(
        res => {
          this.registerSvgSymbolsByString(res);
        },
        () => {
          throw getAuiIconFailedToLoadCustomIconFile(url);
        },
      );
  }

  registerSvgSymbolsByString(str: string) {
    this.appendSvg(str);
  }

  private appendSvg(svgString: string) {
    const setEl = this.doc.createElement('div');
    setEl.className = 'aui-icon-set';
    setEl.style.display = 'none';
    setEl.innerHTML = svgString;
    this.doc.body.append(setEl);
  }
}

// eslint-disable-next-line sonar/function-name
export function ICON_REGISTER_PROVIDER_FACTORY(
  parentRegister: IconRegisterService,
  document: Document,
  httpClient: HttpClient,
) {
  return parentRegister || new IconRegisterService(document, httpClient);
}

export const ICON_REGISTER_SERVICE_PROVIDER = {
  provide: IconRegisterService,
  deps: [
    [new Optional(), new SkipSelf(), IconRegisterService],
    [new Optional(), DOCUMENT],
    [new Optional(), HttpClient],
  ],
  useFactory: ICON_REGISTER_PROVIDER_FACTORY,
};
