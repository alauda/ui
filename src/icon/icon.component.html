<svg
  *ngIf="icon === 'spinner'; else icons"
  [attr.class]="getClass('aui-icon-spinner')"
  [style.fill]="color"
  [style.width]="getWidth()"
  [style.height]="getHeight()"
  viewBox="0 0 200 200"
>
  <defs>
    <linearGradient
      [attr.id]="'right-' + id"
      gradientUnits="userSpaceOnUse"
      x1="150"
      y1="20"
      x2="150"
      y2="180"
    >
      <stop offset="0" style="stop-color: #000" />
      <stop offset="1" style="stop-color: #808080" />
    </linearGradient>
    <linearGradient
      [attr.id]="'left-' + id"
      gradientUnits="userSpaceOnUse"
      x1="50"
      y1="0"
      x2="50"
      y2="180"
    >
      <stop offset="0" style="stop-color: #fff" />
      <stop offset="1" style="stop-color: #808080" />
    </linearGradient>
    <mask [attr.id]="'mask-' + id">
      <g>
        <path
          [attr.fill]="'url(#right-' + id + ')'"
          d="M100,0v20c44.1,0,80,35.9,80,80c0,44.1-35.9,80-80,80v20c55.2,0,100-44.8,100-100S155.2,0,100,0z"
        />
        <path
          [attr.fill]="'url(#left-' + id + ')'"
          d="M20,100c0-44.1,35.9-80,80-80V0C44.8,0,0,44.8,0,100s44.8,100,100,100v-20C55.9,180,20,144.1,20,100z"
        />
      </g>
    </mask>
  </defs>
  <rect
    x="0"
    y="0"
    width="200"
    height="200"
    [attr.mask]="'url(#mask-' + id + ')'"
  />
</svg>

<ng-template #icons>
  <svg
    *ngIf="
      getIconId(
        icon ? icon : ({ light: light, dark: dark } | auiThemePicker)
      ) as iconId
    "
    [attr.class]="getClass(iconId)"
    [style.fill]="color"
    [style.width]="getWidth()"
    [style.height]="getHeight()"
  >
    <use [attr.xlink:href]="link + '#' + iconId"></use>
  </svg>
</ng-template>
