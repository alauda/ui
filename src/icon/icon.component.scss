@import '../theme/var';

aui-icon {
  vertical-align: bottom;
}

.aui-icon {
  display: inline-block;
  width: 1em;
  height: 1em;
  vertical-align: -0.125em;
  fill: currentColor;

  &__background {
    width: 1em;
    height: 1em;
    fill: white;

    &.circle {
      clip-path: circle(45%);
    }

    &.triangle {
      clip-path: polygon(50% 10%, 90% 90%, 10% 90%);
    }
  }

  &--left {
    margin-left: use-var(spacing-s);
  }

  &--right {
    margin-right: use-var(spacing-s);
  }
}

.aui-icon-spinner {
  animation: rotating 0.5s linear infinite;
}

aui-icon + aui-icon > .aui-icon {
  margin-left: use-var(spacing-s);
}

@keyframes rotating {
  0% {
    transform: rotateZ(0deg);
  }
  100% {
    transform: rotateZ(360deg);
  }
}
