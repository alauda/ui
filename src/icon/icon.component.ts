import {
  ChangeDetectionStrategy,
  Component,
  Input,
  ViewEncapsulation,
} from '@angular/core';

import { Bem, buildBem, handlePixel } from '../utils';

import { IconRegisterService } from './icon-register.service';

let id = 0;

@Component({
  selector: 'aui-icon',
  templateUrl: './icon.component.html',
  styleUrls: ['./icon.component.scss'],
  encapsulation: ViewEncapsulation.None,
  changeDetection: ChangeDetectionStrategy.OnPush,
  preserveWhitespaces: false,
})
export class IconComponent {
  @Input()
  icon: string;

  @Input()
  light: string;

  @Input()
  dark: string;

  @Input()
  link = '';

  @Input()
  margin: 'left' | 'right';

  @Input()
  size: string;

  @Input()
  color: string;

  bem: Bem = buildBem('aui-icon');

  id = id++;

  constructor(private readonly iconRegisterService: IconRegisterService) {}

  private parseIcon(icon: string): [string, string] {
    const defaultPrefix = this.iconRegisterService.getDefaultIconPrefix();
    const arr = (icon || '').split(':');
    return arr.length < 2 ? [defaultPrefix, arr[0]] : [arr[0], arr[1]];
  }

  getWidth() {
    if (!this.size) {
      return false;
    }
    return handlePixel((this.size || '').split(',')[0]);
  }

  getHeight() {
    if (!this.size) {
      return false;
    }
    const arr = (this.size || '').split(',');
    return handlePixel(arr[1] || arr[0]);
  }

  getIconId(icon: string) {
    const [prefix, name] = this.parseIcon(icon);
    return `${prefix}-${name}`;
  }

  getClass(id: string) {
    return `${this.bem.block(this.margin)} ${id}`;
  }
}
