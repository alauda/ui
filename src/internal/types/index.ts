export type ValueOf<T> = T[keyof T];

export const ComponentSize = {
  Large: 'large',
  Medium: 'medium',
  Small: 'small',
  Mini: 'mini',
  /**
   * represents default internally, @link https://github.com/angular/vscode-ng-language-service/issues/1147
   */
  Empty: '',
} as const;

export type ComponentSize = ValueOf<typeof ComponentSize>;
