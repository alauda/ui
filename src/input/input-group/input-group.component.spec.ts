import { Component, DebugElement } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';

import { InputModule } from '..';
import { IconModule } from '../../icon';
import { ComponentSize } from '../../types';

describe('InputGroupComponent', () => {
  let fixture: ComponentFixture<TestComponent>;
  let ins: TestComponent;
  let debugEl: DebugElement;
  let el: HTMLElement;
  let groupEl: HTMLElement;
  let innerWrapperEl: HTMLElement;
  let inputWrapperEl: HTMLElement;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [InputModule, IconModule],
      declarations: [TestComponent],
    });
    fixture = TestBed.createComponent(TestComponent);
    fixture.detectChanges();
    ins = fixture.componentInstance;
    debugEl = fixture.debugElement.query(By.css('aui-input-group'));
    el = debugEl.nativeElement;
    groupEl = el.querySelector('.aui-input-group');
    innerWrapperEl = groupEl.querySelector('.aui-input-group__inner-wrapper');
    inputWrapperEl = innerWrapperEl.querySelector(
      '.aui-input-group__input-wrapper',
    );
  });

  it('should render correct component construction', () => {
    expect(groupEl).not.toBeNull();
    expect(groupEl.className).toContain('aui-input-group--medium');
    expect(innerWrapperEl).not.toBeNull();
    expect(
      innerWrapperEl.querySelector('.aui-input-group__input-wrapper'),
    ).not.toBeNull();
  });

  it('should input-wrapper contain aui-input', () => {
    const inputEl = inputWrapperEl.querySelector('input');
    expect(inputEl).not.toBeNull();
    expect(inputEl.className).toContain('aui-input');
    expect(inputEl.className).toContain('aui-input--medium');
  });

  it('should render correct prefix & suffix', () => {
    expect(groupEl.className).toContain('hasPrefix');
    expect(groupEl.className).toContain('hasSuffix');
    expect(
      inputWrapperEl.querySelector('.aui-input-group__prefix'),
    ).not.toBeNull();
    expect(
      inputWrapperEl.querySelector('.aui-input-group__prefix .aui-icon'),
    ).not.toBeNull();
    expect(
      inputWrapperEl.querySelector('.aui-input-group__suffix'),
    ).not.toBeNull();
    expect(
      inputWrapperEl.querySelector('.aui-input-group__suffix .aui-icon'),
    ).not.toBeNull();
  });

  it('should render correct addon', () => {
    const addonEl = innerWrapperEl.querySelectorAll('.aui-input-group__addon');
    expect(addonEl.length).toBe(2);
    expect(addonEl[0]).not.toBeNull();
    expect(addonEl[0].className).toContain('aui-input-group__addon--before');
    expect(addonEl[1]).not.toBeNull();
    expect(addonEl[1].className).toContain('aui-input-group__addon--after');
  });

  it('should size option work correctly', () => {
    ins.size = ComponentSize.Mini;
    fixture.detectChanges();
    expect(groupEl.className).toContain('aui-input-group--mini');
    expect(el.querySelector('input').className).toContain('aui-input--mini');
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});

@Component({
  template: `
    <aui-input-group>
      <span auiInputAddonBefore>HTTPS</span>
      <span auiInputAddonAfter>GB</span>
      <aui-icon
        auiInputPrefix
        icon="search_s"
      ></aui-icon>
      <aui-icon
        auiInputSuffix
        icon="spinner"
      ></aui-icon>
      <input
        aui-input
        [size]="size"
        placeholder="placeholder"
      />
    </aui-input-group>
  `,
})
class TestComponent {
  size: ComponentSize;
}
