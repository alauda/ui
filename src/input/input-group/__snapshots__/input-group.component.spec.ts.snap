// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputGroupComponent should match snapshot 1`] = `
<ng-component>
  <aui-input-group>
    <div
      class="aui-input-group aui-input-group--medium hasAddonBefore hasAddonAfter hasPrefix hasSuffix"
    >
      <div
        class="aui-input-group__inner-wrapper"
      >
        
        <div
          class="aui-input-group__addon aui-input-group__addon--before"
        >
          <span
            auiinputaddonbefore=""
          >
            HTTPS
          </span>
        </div>
        <div
          class="aui-input-group__input-wrapper"
        >
          
          <span
            class="aui-input-group__prefix"
          >
            <aui-icon
              auiinputprefix=""
              icon="search_s"
            >
              
              
              <svg
                class="aui-icon aui-icon-search_s"
              >
                <use
                  xlink:href="#aui-icon-search_s"
                />
              </svg>
              
              
            </aui-icon>
          </span>
          <input
            aui-input=""
            class="aui-input aui-input--medium"
            placeholder="placeholder"
          />
          
          <span
            class="aui-input-group__suffix"
          >
            <aui-icon
              auiinputsuffix=""
              icon="spinner"
            >
              
              <svg
                class="aui-icon aui-icon-spinner"
                viewBox="0 0 200 200"
              >
                <defs>
                  <lineargradient
                    gradientUnits="userSpaceOnUse"
                    id="right-11"
                    x1="150"
                    x2="150"
                    y1="20"
                    y2="180"
                  >
                    <stop
                      offset="0"
                      style="stop-color: #000"
                    />
                    <stop
                      offset="1"
                      style="stop-color: #808080"
                    />
                  </lineargradient>
                  <lineargradient
                    gradientUnits="userSpaceOnUse"
                    id="left-11"
                    x1="50"
                    x2="50"
                    y1="0"
                    y2="180"
                  >
                    <stop
                      offset="0"
                      style="stop-color: #fff"
                    />
                    <stop
                      offset="1"
                      style="stop-color: #808080"
                    />
                  </lineargradient>
                  <mask
                    id="mask-11"
                  >
                    <g>
                      <path
                        d="M100,0v20c44.1,0,80,35.9,80,80c0,44.1-35.9,80-80,80v20c55.2,0,100-44.8,100-100S155.2,0,100,0z"
                        fill="url(#right-11)"
                      />
                      <path
                        d="M20,100c0-44.1,35.9-80,80-80V0C44.8,0,0,44.8,0,100s44.8,100,100,100v-20C55.9,180,20,144.1,20,100z"
                        fill="url(#left-11)"
                      />
                    </g>
                  </mask>
                </defs>
                <rect
                  height="200"
                  mask="url(#mask-11)"
                  width="200"
                  x="0"
                  y="0"
                />
              </svg>
              
            </aui-icon>
          </span>
        </div>
        
        <div
          class="aui-input-group__addon aui-input-group__addon--after"
        >
          <span
            auiinputaddonafter=""
          >
            GB
          </span>
        </div>
      </div>
    </div>
  </aui-input-group>
</ng-component>
`;
